params.go

```diff
@@ -30,6 +30,11 @@ type Params struct {
 	Data interface{} `json:"-"`
 }
 
+func (p *Params) clone() *Params {
+	copy := *p
```

Comment:
you'll need to do a deeper copy here if you want p to be completely separate from copy. Params.Header is a map type, Params.Context, requestParams, and data are interfaces, so they will all be copied by their addresses, not by values.

Copying a map:
```go
for k,v := range originalMap {
  newMap[k] = v
}
```
interfaces are tougher. because you can't copy an interface, since its just a pointer to a struct. 
Let's schedule some time to talk about this